### PR TITLE
Implement ReverseUTF8/Reverse function push down

### PR DESCRIPTION
**Pull request by lizhenhuan**
What problem does this PR solve?
Issue Number: close https://github.com/pingcap/tiflash/issues/5111

Problem Summary: Implement ReverseUTF8/Reverse function push down

What is changed and how it works?
Proposal: [xxx](https://github.com/pingcap/tiflash/pull/url)

What's Changed:

Related https://github.com/pingcap/tidb/pull/35738

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
Support to pushdown reverse to TiFlash
```

### CODE REVIEW

**Comment by Lloyd-Pottiger** [comment]
Thanks for your contribution! Please format the code according to https://github.com/pingcap/tiflash#contributing in order to pass CI build~

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+#pragma GCC diagnostic pop
+
+namespace DB
+{
+namespace tests
+{
+class StringReverse : public DB::tests::FunctionTest
+{
+public:
+    static constexpr auto func_name = "reverse";
```
Comment (mengxin9014) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:41]: Please also test reverseUTF8.

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        return createColumn<Nullable<String>>(v);
+    }
+};
+
+// test string and string
+TEST_F(StringReverse, strAndStrTest)
+try
+{
+    ASSERT_COLUMN_EQ(
+        toVec({"olleh", "dlrow,olleh", "", "pacgnip.www", "。。。试.测.文中"}),
```
Comment (mengxin9014) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:55]: Maybe you can add const and nullable tests, and test strings with other character sets, such as "ЀЁЂЃЄЅІЇЈЉЊЋЌЍЎЏ", "ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩΣ", "ԹՓՁՋՐՉՃԺԾՔՈԵՌՏԸՒԻՕՊԱՍԴՖԳՀՅԿԼԽԶՂՑՎԲՆՄՇ", etc.
You can refer [gtest_strings_upper.test](https://github.com/pingcap/tiflash/blob/25545c452b21a5e545f7d8237364076ed2a057ec/dbms/src/Functions/tests/gtest_strings_upper.cpp#L1)

File: tests/fullstack-test/expr/reverse.test [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+mysql> drop table if exists test.t;
+mysql> create table if not exists test.t(a char(20));
+
+mysql> insert into test.t values('hello');
```
Comment (mengxin9014) [inline tests/fullstack-test/expr/reverse.test:18]: Similarly test strings with other character sets, such as "ЀЁЂЃЄЅІЇЈЉЊЋЌЍЎЏ", "ΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩΣ", "ԹՓՁՋՐՉՃԺԾՔՈԵՌՏԸՒԻՕՊԱՍԴՖԳՀՅԿԼԽԶՂՑՎԲՆՄՇ", etc.

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+        executeFunction(
+            func_name,
+            toVec({"hello", "hello,world", "", "www.pingcap", "中文.测.试。。。"}), ));
+}
+CATCH
+
+// test NULL
+TEST_F(StringLength, nullTest)
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:63]: Why StringLength?

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+#include <Interpreters/Context.h>
+#include <TestUtils/FunctionTestUtils.h>
+#include <TestUtils/TiFlashTestBasic.h>
+
+#include <string>
+#include <vector>
+
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wsign-compare"
+#include <Poco/Types.h>
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:30]: Do we need the header?

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+            "reverseUTF8",
+            toVec(candidate_strings)));
+}
+CATCH
+
+
+// test NULL
+TEST_F(StringReverse, nullTest)
+{
+    ASSERT_COLUMN_EQ(
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:86]: more rows with null and not null maybe better

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <Columns/ColumnString.h>
+#include <Columns/ColumnsNumber.h>
+#include <DataTypes/DataTypeNullable.h>
+#include <DataTypes/DataTypesNumber.h>
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:18]: clear unnecessary headers.

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            toVec(candidate_strings)));
+}
+CATCH
+
+// test reverseUTF8
+TEST_F(StringReverse, stringReverseUTF8Test)
+try
+{
+    std::vector<std::optional<String>> candidate_strings = {"one week's time test", "abc测试def", "abcテストabc", "ѐёђѓєѕіїјљњћќѝўџ", "+ѐ-ё*ђ/ѓ!є@ѕ#і$@ї%ј……љ&њ（ћ）ќ￥ѝ#ў@џ！^", "αβγδεζηθικλμνξοπρστυφχψωσ", "▲α▼βγ➨δε☎ζη✂θι€κλ♫μν✓ξο✚πρ℉στ♥υφ♖χψ♘ω★σ✕", "թփձջրչճժծքոեռտըւիօպասդֆգհյկլխզղցվբնմշ"};
+    std::vector<std::optional<String>> reversed_strings = {"tset emit s'keew eno", "fed试测cba", "cbaトステcba", "џўѝќћњљјїіѕєѓђёѐ", "^！џ@ў#ѝ￥ќ）ћ（њ&љ……ј%ї@$і#ѕ@є!ѓ/ђ*ё-ѐ+", "σωψχφυτσρποξνμλκιθηζεδγβα", "✕σ★ω♘ψχ♖φυ♥τσ℉ρπ✚οξ✓νμ♫λκ€ιθ✂ηζ☎εδ➨γβ▼α▲", "շմնբվցղզխլկյհգֆդսապօիւըտռեոքծժճչրջձփթ"};
```
Comment (ywqzzy) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:68]: why use std::optional<string>?

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+    // ASSERT_COLUMN_EQ(
+    //     toVec(reversed_strings),
+    //     executeFunction(
+    //         func_name,
+    //         toVec(candidate_strings)));
+    ASSERT_COLUMN_EQ(
+        toVec(reversed_strings),
+        executeFunction(
+            "reverseUTF8",
```
Comment (mengxin9014) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:77]: Maybe you can define a variable for "reverseUTF8" or use "reverse" directly for testing reverse function, just be consistent.

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+}
+CATCH
+
+// test reverseUTF8
+TEST_F(StringReverse, stringReverseUTF8Test)
+try
+{
+    std::vector<std::optional<String>> candidate_strings = {"one week's time test", "abc测试def", "abcテストabc", "ѐёђѓєѕіїјљњћќѝўџ", "+ѐ-ё*ђ/ѓ!є@ѕ#і$@ї%ј……љ&њ（ћ）ќ￥ѝ#ў@џ！^", "αβγδεζηθικλμνξοπρστυφχψωσ", "▲α▼βγ➨δε☎ζη✂θι€κλ♫μν✓ξο✚πρ℉στ♥υφ♖χψ♘ω★σ✕", "թփձջրչճժծքոեռտըւիօպասդֆգհյկլխզղցվբնմշ"};
+    std::vector<std::optional<String>> reversed_strings = {"tset emit s'keew eno", "fed试测cba", "cbaトステcba", "џўѝќћњљјїіѕєѓђёѐ", "^！џ@ў#ѝ￥ќ）ћ（њ&љ……ј%ї@$і#ѕ@є!ѓ/ђ*ё-ѐ+", "σωψχφυτσρποξνμλκιθηζεδγβα", "✕σ★ω♘ψχ♖φυ♥τσ℉ρπ✚οξ✓νμ♫λκ€ιθ✂ηζ☎εδ➨γβ▼α▲", "շմնբվցղզխլկյհգֆդսապօիւըտռեոքծժճչրջձփթ"};
+    // ASSERT_COLUMN_EQ(
```
Comment (mengxin9014) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:69]: Remove unnecessary code.

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+{
+namespace tests
+{
+class StringReverse : public DB::tests::FunctionTest
+{
+public:
+    static constexpr auto func_name = "reverse";
+
+protected:
+    ColumnWithTypeAndName toVec(const std::vector<std::optional<String>> & v)
```
Comment (mengxin9014) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:44]: Testing nullable means that the column can be null, you can refer here https://github.com/pingcap/tiflash/blob/25545c452b21a5e545f7d8237364076ed2a057ec/dbms/src/Functions/tests/gtest_strings_upper.cpp#L41, https://github.com/pingcap/tiflash/blob/25545c452b21a5e545f7d8237364076ed2a057ec/dbms/src/Functions/tests/gtest_strings_upper.cpp#L59

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp (hunk 9 above)
Comment (lizhenhuan) [reply dbms/src/Functions/tests/gtest_strings_reverse.cpp:67]: I change all these to constant

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp (hunk 10 above)
Comment (lizhenhuan) [reply dbms/src/Functions/tests/gtest_strings_reverse.cpp]: removed

File: dbms/src/Functions/tests/gtest_strings_reverse.cpp [hunk 12]
```diff
@@ -0,0 +1,104 @@
...
+        return createColumn<Nullable<String>>(v);
+    }
+
+    static ColumnWithTypeAndName toConst(const String & s)
+    {
+        return createConstColumn<String>(1, s);
+    }
+};
+// test reverse
+TEST_F(StringReverse, stringReverseTest)
```
Comment (mengxin9014) [inline dbms/src/Functions/tests/gtest_strings_reverse.cpp:47]: Maybe you can refactor the TEST function like:
```cpp
// test reverse
TEST_F(StringReverse, stringReverseTest)
try
{
    // test vector
    code...
    // test nullable
    code...
    // test const
    code...
    // test null
    code...
}
CATCH

// test reverseUTF8
TEST_F(StringReverse, stringReverseUTF8Test)
try
{
    // test vector
    code...
    // test nullable
    code...
    // test const
    code...
    // test null
    code...
}
CATCH
```
This makes the test code structure more clearer

**Comment by mengxin9014** [comment]
LGTM

**Review by mengxin9014 — APPROVED** [review]

**Review by ywqzzy — APPROVED** [review]
LGTM! Nice job!

**Review by LittleFall — APPROVED** [review]

**Pull request merged**